src/poetry/packages/locker.py

```diff
@@ -203,145 +204,111 @@ def locked_repository(self, with_dev_reqs: bool = False) -> Repository:
 
-                if not pinned_versions:
-                    requirement.set_constraint(constraint)
+            locked_package = cls.__get_locked_package(
+                requirement, packages_by_name, nested_dependencies
+            )
 
-                for require in locked_package.requires:
-                    if require.marker.is_empty():
-                        require.marker = requirement.marker
-                    else:
-                        require.marker = require.marker.intersect(requirement.marker)
+            if not locked_package:
+                raise RuntimeError(f"Dependency walk failed at {requirement}")
 
-                    require.marker = require.marker.intersect(locked_package.marker)
+            # create dependency from locked package to retain dependency metadata
+            # if this is not done, we can end-up with incorrect nested dependencies
+            constraint = requirement.constraint
+            marker = requirement.marker
+            extras = requirement.extras
+            requirement = locked_package.to_dependency()
+            requirement.marker = requirement.marker.intersect(marker)
 
-                    if key not in nested_dependencies:
-                        next_level_dependencies.append(require)
+            requirement.set_constraint(constraint)
 
-            if requirement.name in project_level_dependencies and level == 0:
-                # project level dependencies take precedence
-                continue
+            for require in locked_package.requires:
+                if require.in_extras and extras.isdisjoint(require.in_extras):
+                    continue
 
-            if not locked_package:
-                # we make a copy to avoid any side-effects
-                requirement = deepcopy(requirement)
+                require = deepcopy(require)
+                require.marker = require.marker.intersect(requirement.marker)
+                if not require.marker.is_empty():
+                    dependencies.append(require)
 
+            key = locked_package
             if key not in nested_dependencies:
                 nested_dependencies[key] = requirement
             else:
                 nested_dependencies[key].marker = nested_dependencies[key].marker.union(
                     requirement.marker
                 )
 
-        return cls.__walk_dependency_level(
-            dependencies=next_level_dependencies,
-            level=level + 1,
-            pinned_versions=pinned_versions,
-            packages_by_name=packages_by_name,
-            project_level_dependencies=project_level_dependencies,
-            nested_dependencies=nested_dependencies,
-        )
+        return nested_dependencies
 
     @classmethod
     def get_project_dependencies(
         cls,
         project_requires: list[Dependency],
         locked_packages: list[Package],
-        pinned_versions: bool = False,
-        with_nested: bool = False,
```

Comment:
Is there a reason why we are droping this functionality? Export should not, in all cases provide locked versions nor nested dependencies. With this change, only the `pip freeze` use case can be addressed.

Reply:
it was unused function, no?  I removed it because I found no callers using it.

Reply:
There is an intention to move Locker logic into core to support use cases where you would need pinned versions in wheels. This can be either for direct dependencies or nested. That was the rationale why this was introduced. I would suggest we keep this ability in so we can move it over easier later.

Reply:
This function is currently both unused and untested.  And to be honest I'm not sure exactly what those parameters even mean.

I'd much prefer to get the case that we have working; we can always add function - with testcases! - later if and when it is wanted or needed.

Reply:
In general, irrespective of test coverage, I would be typically reluctant to change the signature of a public class method. Especially when refactoring, unlesss there is a pressing need. However, this is probably a special scenario as this is only used by `poetry` and `poetry-export-plugin` at present.

For some context this, I beleive, the parameters were introduced as as step change while working on the following:

- https://github.com/python-poetry/poetry-core/pull/83
- https://github.com/python-poetry/poetry-core/pull/71

See the example metadata in 71 to see the originally intended use cases.

For one reason or the other, progresson that halted but I do intend to pick it up again.

Additionally, export plugin will also eventualy need 2 kinds of exports, locked and constrainted exports.

In retrospect, you are correct in that more test coverage should be present. However, I would prefer coverage be added rather than functionality removed. That said, in this scenario I do think either is fine as long as the new changes do not somehow preclude the re-introduction of the functionality.

I'll go over the changes with a fresh pair of eyes today to see if applying this breaks something or makes things harder to change once the exoirt plugin gets decoupled.

Fwiw, apprecate the changes you have put into this, this is really useful for improving the maintainability, :pray: